### ISSUE TITLE
--stream-segment-threads option doesn't seem to be working with HLS streams

### ISSUE THREAD

**Issue by kikuyan** [streamlink/streamlink#4126]
### Checklist

- [X] This is a bug report and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed bug reports](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22bug%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest build from the master branch

### Description

`--stream-segment-threads` option doesn't seem to be working with HLS streams.
I can't see increase in download speed and segments do not seem to be downloaded in parallel.
Also, using the `--stream-segment-threads` option seems to have a tendency to fail to download (fail frequently, however, the dailymotion sample may be bad).

3 logs for comparison below:
<span>1)</span> 2.4.0 without `--stream-segment-threads`, 
2) 2.4.0 with `--stream-segment-threads 4`, and 
3) 2.3.0 with `--hls-segment-threads 4`

### Debug log

```text
1) 2.4.0 without --stream-segment-threads
streamlink -l debug --hls-duration 01:00 https://www.dailymotion.com/video/x853fvr 720p -o /dev/null

[cli][debug] OS:         macOS 10.12.6
[cli][debug] Python:     3.9.7
[cli][debug] Streamlink: 2.4.0+62.g501e4a0
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.2.1)
[cli][debug] Arguments:
[cli][debug]  url=https://www.dailymotion.com/video/x853fvr
[cli][debug]  stream=['720p']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=/dev/null
[cli][debug]  --hls-duration=60
[cli][info] Found matching plugin dailymotion for URL https://www.dailymotion.com/video/x853fvr
[plugins.dailymotion][debug] Found media ID: x853fvr
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 112p_alt (worst), 112p, 184p_alt, 184p, 288p_alt, 288p, 480p_alt, 480p, 720p_alt, 720p, 1080p_alt, 1080p (best)
[cli][info] Opening stream: 720p (hls)
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 0; Last Sequence: 53
[stream.hls][debug] Start offset: 0; Duration: 60; Start Sequence: 0; End Sequence: 53
[stream.hls][debug] Adding segment 0 to queue
[stream.hls][debug] Adding segment 1 to queue
[stream.hls][debug] Adding segment 2 to queue
[stream.hls][debug] Adding segment 3 to queue
[stream.hls][debug] Adding segment 4 to queue
[stream.hls][debug] Adding segment 5 to queue
[stream.hls][debug] Adding segment 6 to queue
[stream.hls][debug] Adding segment 7 to queue
[stream.hls][debug] Adding segment 8 to queue
[stream.hls][debug] Adding segment 9 to queue
[stream.hls][debug] Adding segment 10 to queue
[stream.hls][debug] Adding segment 11 to queue
[stream.hls][debug] Adding segment 12 to queue
[stream.hls][debug] Adding segment 13 to queue
[stream.hls][debug] Adding segment 14 to queue
[stream.hls][debug] Adding segment 15 to queue
[stream.hls][debug] Adding segment 16 to queue
[stream.hls][debug] Adding segment 17 to queue
[stream.hls][debug] Adding segment 18 to queue
[stream.hls][debug] Adding segment 19 to queue
[stream.hls][info] Stopping stream early after 60
[stream.segmented][debug] Closing worker thread
[cli][debug] Checking file output
[cli][debug] Writing stream to output
[download][null] Written 456.0 KB (0s @ 833.6 KB/s)                             [stream.hls][debug] Segment 0 complete
[download][null] Written 1.0 MB (1s @ 965.1 KB/s)                               [stream.hls][debug] Segment 1 complete
[download][null] Written 1.9 MB (2s @ 943.5 KB/s)                               [stream.hls][debug] Segment 2 complete
[download][null] Written 2.9 MB (3s @ 965.8 KB/s)                               [stream.hls][debug] Segment 3 complete
[download][null] Written 3.7 MB (3s @ 1.0 MB/s)                                 [stream.hls][debug] Segment 4 complete
[download][null] Written 4.6 MB (4s @ 1.0 MB/s)                                 [stream.hls][debug] Segment 5 complete
[download][null] Written 5.5 MB (5s @ 1.0 MB/s)                                 [stream.hls][debug] Segment 6 complete
[download][null] Written 6.4 MB (6s @ 893.6 KB/s)                               [stream.hls][debug] Segment 7 complete
[download][null] Written 7.0 MB (7s @ 986.4 KB/s)                               [stream.hls][debug] Segment 8 complete
[download][null] Written 7.9 MB (8s @ 972.8 KB/s)                               [stream.hls][debug] Segment 9 complete
[download][null] Written 8.4 MB (8s @ 969.7 KB/s)                               [stream.hls][debug] Segment 10 complete
[download][null] Written 9.3 MB (9s @ 929.2 KB/s)                               [stream.hls][debug] Segment 11 complete
[download][null] Written 9.9 MB (10s @ 954.8 KB/s)                              [stream.hls][debug] Segment 12 complete
[download][null] Written 10.9 MB (11s @ 972.7 KB/s)                             [stream.hls][debug] Segment 13 complete
[download][null] Written 11.8 MB (12s @ 973.1 KB/s)                             [stream.hls][debug] Segment 14 complete
[download][null] Written 12.8 MB (13s @ 753.6 KB/s)                             [stream.hls][debug] Segment 15 complete
[download][null] Written 13.4 MB (14s @ 646.1 KB/s)                             [stream.hls][debug] Segment 16 complete
[download][null] Written 14.5 MB (16s @ 655.7 KB/s)                             [stream.hls][debug] Segment 17 complete
[download][null] Written 15.2 MB (17s @ 672.8 KB/s)                             [stream.hls][debug] Segment 18 complete
[download][null] Written 15.9 MB (18s @ 671.1 KB/s)                             [stream.hls][debug] Segment 19 complete
[stream.segmented][debug] Closing writer thread

[cli][info] Stream ended
[cli][info] Closing currently open stream...


2) 2.4.0 with --stream-segment-threads 4
streamlink -l debug --hls-duration 01:00 --stream-segment-threads 4 https://www.dailymotion.com/video/x853fvr 720p -o /dev/null

[cli][debug] OS:         macOS 10.12.6
[cli][debug] Python:     3.9.7
[cli][debug] Streamlink: 2.4.0+62.g501e4a0
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.2.1)
[cli][debug] Arguments:
[cli][debug]  url=https://www.dailymotion.com/video/x853fvr
[cli][debug]  stream=['720p']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=/dev/null
[cli][debug]  --stream-segment-threads=4
[cli][debug]  --hls-duration=60
[cli][info] Found matching plugin dailymotion for URL https://www.dailymotion.com/video/x853fvr
[plugins.dailymotion][debug] Found media ID: x853fvr
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 112p_alt (worst), 112p, 184p_alt, 184p, 288p_alt, 288p, 480p_alt, 480p, 720p_alt, 720p, 1080p_alt, 1080p (best)
[cli][info] Opening stream: 720p (hls)
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 0; Last Sequence: 53
[stream.hls][debug] Start offset: 0; Duration: 60; Start Sequence: 0; End Sequence: 53
[stream.hls][debug] Adding segment 0 to queue
[stream.hls][debug] Adding segment 1 to queue
[stream.hls][debug] Adding segment 2 to queue
[stream.hls][debug] Adding segment 3 to queue
[stream.hls][debug] Adding segment 4 to queue
[stream.hls][debug] Adding segment 5 to queue
[stream.hls][debug] Adding segment 6 to queue
[stream.hls][debug] Adding segment 7 to queue
[stream.hls][debug] Adding segment 8 to queue
[stream.hls][debug] Adding segment 9 to queue
[stream.hls][debug] Adding segment 10 to queue
[stream.hls][debug] Adding segment 11 to queue
[stream.hls][debug] Adding segment 12 to queue
[stream.hls][debug] Adding segment 13 to queue
[stream.hls][debug] Adding segment 14 to queue
[stream.hls][debug] Adding segment 15 to queue
[stream.hls][debug] Adding segment 16 to queue
[stream.hls][debug] Adding segment 17 to queue
[stream.hls][debug] Adding segment 18 to queue
[stream.hls][debug] Adding segment 19 to queue
[stream.hls][info] Stopping stream early after 60
[stream.segmented][debug] Closing worker thread
[cli][debug] Checking file output
[cli][debug] Writing stream to output
[download][null] Written 456.0 KB (0s @ 845.6 KB/s)                             [stream.hls][debug] Segment 0 complete
[download][null] Written 1011.8 KB (1s @ 968.0 KB/s)                            [stream.hls][debug] Segment 1 complete
[download][null] Written 1.9 MB (2s @ 927.0 KB/s)                               [stream.hls][debug] Segment 2 complete
[download][null] Written 2.9 MB (3s @ 941.8 KB/s)                               [stream.hls][debug] Segment 3 complete
[download][null] Written 3.7 MB (4s @ 872.0 KB/s)                               [stream.hls][debug] Segment 4 complete
[download][null] Written 4.7 MB (5s @ 912.8 KB/s)                               [stream.hls][debug] Segment 5 complete
[download][null] Written 5.3 MB (5s @ 946.4 KB/s)                               [stream.hls][debug] Segment 6 complete
[download][null] Written 6.2 MB (6s @ 976.1 KB/s)                               [stream.hls][debug] Segment 7 complete
[download][null] Written 7.1 MB (7s @ 927.1 KB/s)                               [stream.hls][debug] Segment 8 complete
[download][null] Written 7.9 MB (8s @ 838.4 KB/s)                               [stream.hls][debug] Segment 9 complete
[download][null] Written 8.5 MB (9s @ 890.3 KB/s)                               [stream.hls][debug] Segment 10 complete
[download][null] Written 9.3 MB (10s @ 890.8 KB/s)                              [stream.hls][debug] Segment 11 complete
[download][null] Written 10.2 MB (11s @ 894.6 KB/s)                             [stream.hls][debug] Segment 12 complete
[download][null] Written 10.6 MB (11s @ 870.2 KB/s)                             [stream.hls][debug] Segment 13 complete
[download][null] Written 11.9 MB (13s @ 778.8 KB/s)                             [stream.hls][debug] Segment 14 complete
[download][null] Written 12.6 MB (14s @ 720.3 KB/s)                             [stream.hls][debug] Segment 15 complete
[download][null] Written 13.7 MB (16s @ 628.4 KB/s)                             [stream.hls][debug] Segment 16 complete
[stream.hls][error] Download of segment 17 failed (('Connection broken: IncompleteRead(319360 bytes read, 559352 more expected)', IncompleteRead(319360 bytes read, 559352 more expected)))
[download][null] Written 14.2 MB (17s @ 641.7 KB/s)                             [stream.hls][error] Download of segment 18 failed (('Connection broken: IncompleteRead(275124 bytes read, 600392 more expected)', IncompleteRead(275124 bytes read, 600392 more expected)))
[download][null] Written 14.5 MB (17s @ 606.0 KB/s)                             [stream.hls][error] Download of segment 19 failed (('Connection broken: IncompleteRead(275124 bytes read, 453564 more expected)', IncompleteRead(275124 bytes read, 453564 more expected)))
[stream.segmented][debug] Closing writer thread

[cli][info] Stream ended
[cli][info] Closing currently open stream...


3) 2.3.0 with --hls-segment-threads 4
streamlink -l debug --hls-duration 01:00 --hls-segment-threads 4 https://www.dailymotion.com/video/x853fvr 720p -o /dev/null

[cli][debug] OS:         macOS 10.12.6
[cli][debug] Python:     3.9.7
[cli][debug] Streamlink: 2.3.0
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.2.1)
[cli][debug] Arguments:
[cli][debug]  url=https://www.dailymotion.com/video/x853fvr
[cli][debug]  stream=['720p']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=/dev/null
[cli][debug]  --hls-segment-threads=4
[cli][debug]  --hls-duration=60
[cli][info] Found matching plugin dailymotion for URL https://www.dailymotion.com/video/x853fvr
[plugins.dailymotion][debug] Found media ID: x853fvr
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 112p_alt (worst), 112p, 184p_alt, 184p, 288p_alt, 288p, 480p_alt, 480p, 720p_alt, 720p, 1080p_alt, 1080p (best)
[cli][info] Opening stream: 720p (hls)
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 0; Last Sequence: 53
[stream.hls][debug] Start offset: 0; Duration: 60; Start Sequence: 0; End Sequence: 53
[stream.hls][debug] Adding segment 0 to queue
[stream.hls][debug] Adding segment 1 to queue
[stream.hls][debug] Adding segment 2 to queue
[stream.hls][debug] Adding segment 3 to queue
[stream.hls][debug] Adding segment 4 to queue
[stream.hls][debug] Adding segment 5 to queue
[stream.hls][debug] Adding segment 6 to queue
[stream.hls][debug] Adding segment 7 to queue
[stream.hls][debug] Adding segment 8 to queue
[stream.hls][debug] Adding segment 9 to queue
[stream.hls][debug] Adding segment 10 to queue
[stream.hls][debug] Adding segment 11 to queue
[stream.hls][debug] Adding segment 12 to queue
[stream.hls][debug] Adding segment 13 to queue
[stream.hls][debug] Adding segment 14 to queue
[stream.hls][debug] Adding segment 15 to queue
[stream.hls][debug] Adding segment 16 to queue
[stream.hls][debug] Adding segment 17 to queue
[stream.hls][debug] Adding segment 18 to queue
[stream.hls][debug] Adding segment 19 to queue
[stream.hls][info] Stopping stream early after 60
[stream.segmented][debug] Closing worker thread
[stream.hls][debug] Segment 0 complete
[cli][debug] Checking file output
[cli][debug] Writing stream to output
[stream.hls][debug] Segment 1 complete
[stream.hls][debug] Segment 2 complete
[stream.hls][debug] Segment 3 complete
[stream.hls][debug] Segment 4 complete
[download][null] Written 3.0 MB (1s @ 2.5 MB/s)                                 [stream.hls][debug] Segment 5 complete
[stream.hls][debug] Segment 6 complete
[stream.hls][debug] Segment 7 complete
[stream.hls][debug] Segment 8 complete
[download][null] Written 6.5 MB (1s @ 3.3 MB/s)                                 [stream.hls][debug] Segment 9 complete
[stream.hls][debug] Segment 10 complete
[stream.hls][debug] Segment 11 complete
[download][null] Written 8.6 MB (2s @ 3.4 MB/s)                                 [stream.hls][debug] Segment 12 complete
[stream.hls][debug] Segment 13 complete
[stream.hls][debug] Segment 14 complete
[stream.hls][debug] Segment 15 complete
[download][null] Written 11.1 MB (3s @ 3.0 MB/s)                                [stream.hls][debug] Segment 16 complete
[stream.hls][debug] Segment 17 complete
[stream.hls][debug] Segment 18 complete
[download][null] Written 14.5 MB (4s @ 3.2 MB/s)                                [stream.hls][debug] Segment 19 complete
[stream.segmented][debug] Closing writer thread

[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

**Comment by kikuyan** [comment]
While making this issue report, I noticed that streamlink 2.4.0 opens many tcp connections at the same time (almost as many as segment queue entries).
Is this the intended behavior? But even so, I think this consumes much resources unnecessarily and need to be improved.

Here are 2 excerpts of `lsof` output: `lsof -c Python -a -i4:https -l -r2` while running `streamlink -l debug --hls-duration 01:00 https://www.dailymotion.com/video/x853fvr 720p -o /dev/null` (2.4.0 and 2.3.0; same OS, Python version, python modules as above.)

```
1) lsof while running 2.4.0

=======
COMMAND   PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
Python  13980      501    6u  IPv4 0x8054432442599be1      0t0  TCP 192.168.10.101:54658->157.14.84.34.bc.googleusercontent.com:https (ESTABLISHED)
Python  13980      501    7u  IPv4 0x805443243fa928b9      0t0  TCP 192.168.10.101:54659->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501    8u  IPv4 0x80544324425b31b1      0t0  TCP 192.168.10.101:54660->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501    9u  IPv4 0x80544324425801b1      0t0  TCP 192.168.10.101:54661->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   11u  IPv4 0x805443244257cbe1      0t0  TCP 192.168.10.101:54662->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   12u  IPv4 0x80544324425b28b9      0t0  TCP 192.168.10.101:54663->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   13u  IPv4 0x805443244259a4d9      0t0  TCP 192.168.10.101:54664->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   14u  IPv4 0x80544324425b0dd1      0t0  TCP 192.168.10.101:54665->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   15u  IPv4 0x80544324425b16c9      0t0  TCP 192.168.10.101:54666->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   16u  IPv4 0x80544324425afbe1      0t0  TCP 192.168.10.101:54667->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   17u  IPv4 0x80544324425b1fc1      0t0  TCP 192.168.10.101:54668->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   18u  IPv4 0x80544324425b04d9      0t0  TCP 192.168.10.101:54669->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   19u  IPv4 0x8054432433f176c9      0t0  TCP 192.168.10.101:54670->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   20u  IPv4 0x8054432433f17fc1      0t0  TCP 192.168.10.101:54671->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   21u  IPv4 0x8054432433f15be1      0t0  TCP 192.168.10.101:54672->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  13980      501   22u  IPv4 0x8054432433f188b9      0t0  TCP 192.168.10.101:54673->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
=======


2) lsof while running 2.3.0

=======
COMMAND   PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
Python  13997      501    6u  IPv4 0x8054432433f188b9      0t0  TCP 192.168.10.101:54675->157.14.84.34.bc.googleusercontent.com:https (ESTABLISHED)
Python  13997      501    7u  IPv4 0x8054432433f15be1      0t0  TCP 192.168.10.101:54676->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
=======
```

**Comment by bastimeyer** [comment]
> `--stream-segment-threads` option doesn't seem to be working with HLS streams.

You can set a breakpoint in the [segmented writer thread's constructor](https://github.com/streamlink/streamlink/blob/bf269e211a725b44448eb0c2bf84104230327d80/src/streamlink/stream/segmented.py#L101-L122) to see that it's creating the right amount of threads in the segmented writer threadpool executor. The change of the parameter names (#3893) is working correctly, and if there are any issues, it's not because of that.

> I can't see increase in download speed and segments do not seem to be downloaded in parallel.

The segmented writer threadpool will [fetch segments](https://github.com/streamlink/streamlink/blob/bf269e211a725b44448eb0c2bf84104230327d80/src/streamlink/stream/segmented.py#L143) as long as segments are available in the queue, which gets populated by the HLS worker thread that parses the HLS playlist. If the playlist doesn't have more segments available, then there's nothing to queue.

The [fetch logic on the HLS implementation](https://github.com/streamlink/streamlink/blob/bf269e211a725b44448eb0c2bf84104230327d80/src/streamlink/stream/hls.py#L128-L153) consists of simple HTTP requests made by python-requests. There is no control here over how many TCP connections are opened. This depends on how the python-requests Session handles keep-alive connections for various different requests.

[The python-requests documentation](https://docs.python-requests.org/en/master/user/advanced/#keep-alive) says that those connections can only be freed from the Session's connections pool when the buffer has been emptied on `stream=True` requests. `stream=True` is the case nowadays for all HLS segment requests, so that the output doesn't have to wait until the entire segment has completed.

The [`write`](https://github.com/streamlink/streamlink/blob/bf269e211a725b44448eb0c2bf84104230327d80/src/streamlink/stream/hls.py#L173-L204) call which empties the request's buffer happens [as soon as possible](https://github.com/streamlink/streamlink/blob/bf269e211a725b44448eb0c2bf84104230327d80/src/streamlink/stream/segmented.py#L180-L204), depending on the order of completed segments and the availablity of space on Streamlink's ringbuffer. If Streamlink's ringbuffer is full, the write call must halt until there's enough space available again on the buffer.

> Also, using the `--stream-segment-threads` option seems to have a tendency to fail to download (fail frequently, however, the dailymotion sample may be bad).

A segment failure is the result of an error response or connection error which can't be recovered from by re-requesting the segment. HTTP requests of any kind will always be re-attempted by a certain number with a debounce time in Streamlink's HTTPSession subclass.

----

What happens if you disable `stream=True` by setting the second param to `False` here:
https://github.com/streamlink/streamlink/blob/bf269e211a725b44448eb0c2bf84104230327d80/src/streamlink/stream/hls.py#L130

**Comment by kikuyan** [comment]
Thank you, but I can't fully understand your in-depth explanation.

> What happens if you disable stream=True by setting the second param to False here:
streamlink/src/streamlink/stream/hls.py
Line 130 in bf269e2
 return self._fetch(sequence.segment, not sequence.segment.key) 

Works good. And all the problems I was concerned about seem to be resolved.

```
0) git diff
diff --git a/src/streamlink/stream/hls.py b/src/streamlink/stream/hls.py
index e4be9fa1..5a0a95fc 100644
--- a/src/streamlink/stream/hls.py
+++ b/src/streamlink/stream/hls.py
@@ -127,7 +127,7 @@ class HLSStreamWriter(SegmentedStreamWriter):
 
     def fetch(self, sequence: Sequence) -> Optional[Response]:
         try:
-            return self._fetch(sequence.segment, not sequence.segment.key)
+            return self._fetch(sequence.segment, False)
         except StreamError as err:  # pragma: no cover
             log.error(f"Failed to fetch segment {sequence.num}: {err}")
 

1) streamlink -l debug --hls-duration 01:00 --stream-segment-threads 4 https://www.dailymotion.com/video/x853fvr 720p -o /dev/null
[cli][debug] OS:         macOS 10.12.6
[cli][debug] Python:     3.9.7
[cli][debug] Streamlink: 2.4.0+62.g501e4a0.dirty
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.2.1)
[cli][debug] Arguments:
[cli][debug]  url=https://www.dailymotion.com/video/x853fvr
[cli][debug]  stream=['720p']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=/dev/null
[cli][debug]  --stream-segment-threads=4
[cli][debug]  --hls-duration=60
[cli][info] Found matching plugin dailymotion for URL https://www.dailymotion.com/video/x853fvr
[plugins.dailymotion][debug] Found media ID: x853fvr
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 112p_alt (worst), 112p, 184p_alt, 184p, 288p_alt, 288p, 480p_alt, 480p, 720p_alt, 720p, 1080p_alt, 1080p (best)
[cli][info] Opening stream: 720p (hls)
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 0; Last Sequence: 53
[stream.hls][debug] Start offset: 0; Duration: 60; Start Sequence: 0; End Sequence: 53
[stream.hls][debug] Adding segment 0 to queue
[stream.hls][debug] Adding segment 1 to queue
[stream.hls][debug] Adding segment 2 to queue
[stream.hls][debug] Adding segment 3 to queue
[stream.hls][debug] Adding segment 4 to queue
[stream.hls][debug] Adding segment 5 to queue
[stream.hls][debug] Adding segment 6 to queue
[stream.hls][debug] Adding segment 7 to queue
[stream.hls][debug] Adding segment 8 to queue
[stream.hls][debug] Adding segment 9 to queue
[stream.hls][debug] Adding segment 10 to queue
[stream.hls][debug] Adding segment 11 to queue
[stream.hls][debug] Adding segment 12 to queue
[stream.hls][debug] Adding segment 13 to queue
[stream.hls][debug] Adding segment 14 to queue
[stream.hls][debug] Adding segment 15 to queue
[stream.hls][debug] Adding segment 16 to queue
[stream.hls][debug] Adding segment 17 to queue
[stream.hls][debug] Adding segment 18 to queue
[stream.hls][debug] Adding segment 19 to queue
[stream.hls][info] Stopping stream early after 60
[stream.segmented][debug] Closing worker thread
[stream.hls][debug] Segment 0 complete
[cli][debug] Checking file output
[cli][debug] Writing stream to output
[stream.hls][debug] Segment 1 complete
[stream.hls][debug] Segment 2 complete
[stream.hls][debug] Segment 3 complete
[stream.hls][debug] Segment 4 complete
[download][null] Written 3.0 MB (0s @ 3.0 MB/s)                                 [stream.hls][debug] Segment 5 complete
[stream.hls][debug] Segment 6 complete
[stream.hls][debug] Segment 7 complete
[stream.hls][debug] Segment 8 complete
[download][null] Written 6.5 MB (1s @ 3.7 MB/s)                                 [stream.hls][debug] Segment 9 complete
[stream.hls][debug] Segment 10 complete
[stream.hls][debug] Segment 11 complete
[download][null] Written 8.6 MB (2s @ 3.7 MB/s)                                 [stream.hls][debug] Segment 12 complete
[download][null] Written 9.4 MB (2s @ 3.3 MB/s)                                 [stream.hls][debug] Segment 13 complete
[stream.hls][debug] Segment 14 complete
[stream.hls][debug] Segment 15 complete
[download][null] Written 11.1 MB (3s @ 3.0 MB/s)                                [stream.hls][debug] Segment 16 complete
[stream.hls][debug] Segment 17 complete
[stream.hls][debug] Segment 18 complete
[download][null] Written 14.5 MB (4s @ 3.3 MB/s)                                [stream.hls][debug] Segment 19 complete
[stream.segmented][debug] Closing writer thread

[cli][info] Stream ended
[cli][info] Closing currently open stream...


2) lsof -c Python -a -i4:https -l -r2
=======
COMMAND   PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
Python  15856      501    6u  IPv4 0x8054432441ec04d9      0t0  TCP 192.168.10.101:55740->157.14.84.34.bc.googleusercontent.com:https (ESTABLISHED)
Python  15856      501    7u  IPv4 0x80544324452b3fc1      0t0  TCP 192.168.10.101:55741->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  15856      501    8u  IPv4 0x80544324452b1be1      0t0  TCP 192.168.10.101:55743->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  15856      501    9u  IPv4 0x80544324452b2dd1      0t0  TCP 192.168.10.101:55744->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
Python  15856      501   11u  IPv4 0x80544324452b51b1      0t0  TCP 192.168.10.101:55742->proxy-04.sg1.dailymotion.com:https (ESTABLISHED)
=======
```

**Comment by bastimeyer** [comment]
That might be a bug/issue in requests's HTTPAdapter or Response class, or urllib3's PoolManager.

**Comment by kikuyan** [comment]
Wherever the cause resides, I hope that some workaround will be taken in the next release. To us users, isn't it a function degradation in version 2.4.0?

**Comment by bastimeyer** [comment]
This has to be properly debugged and evaluated first, which isn't particularly simple here, especially when it's a potential issue two dependencies deep. I haven't had the time yet to take a deeper look at this as I'm busy with other stuff which I want to finish first. As always, any help from other contributors is appreciated.

I've already explained earlier that segment requests with `stream=True` enable Streamlink to output the segment contents before the entire request has completed, which is important for low latency streaming. The `stream=True` parameter became the default after `--hls-segment-stream-data` was removed in #3894. This is completely unrelated to #3893 which deprecated `--hls-segment-threads`.

According to the python-requests docs, keep-alive connections of HTTP requests with the `stream=True` parameter should be made available in the connection pool again once the entire content was read, but that doesn't seem to work properly from what it looks like.

I don't think that the way Streamlink is making the HTTP requests with `stream=True` and reading the content via `response.iter_content(chunk_size)` is wrong, which means that if this is a bug in one of the dependencies (python-requests / urllib3), then this has to be isolated and demonstrated in a reproducible bug report.
